src/blob/index/bptree/core.rs

```diff
@@ -0,0 +1,336 @@
+                // We used get mut instead of entry(..).or_insert(..) because in second case we
+                // need to clone key everytime. Whereas in first case - only if we insert new
+                // entry.
+                if let Some(v) = headers.get_mut(header.key()) {
+                    v.push(header)
+                } else {
+                    headers.insert(header.key().to_vec(), vec![header]);
+                }
+                Ok(headers)
+            })
+            .map(|headers| (headers, self.header.records_count))
+    }
+
+    async fn get_any(&self, key: &[u8]) -> Result<Option<RecordHeader>> {
+        let root_offset = self.metadata.tree_offset;
+        let mut buf = [0u8; BLOCK_SIZE];
+        let leaf_offset = self.find_leaf_node(key, root_offset, &mut buf).await?;
+        self.read_header(leaf_offset, key, &mut buf).await
+    }
+
+    fn validate(&self) -> Result<()> {
+        if self.header.is_written() {
+            Ok(())
+        } else {
+            Err(Error::validation("Index Header is not valid").into())
+        }
+    }
+}
+
+impl BPTreeFileIndex {
+    async fn find_leaf_node(&self, key: &[u8], mut offset: u64, buf: &mut [u8]) -> Result<u64> {
+        while offset < self.metadata.leaves_offset {
+            offset = if offset == self.metadata.tree_offset {
+                Node::key_offset_serialized(&self.root_node, key)?
+            } else {
+                self.file.read_at(buf, offset).await?;
+                Node::key_offset_serialized(buf, key)?
+            };
+        }
+        Ok(offset)
+    }
+
+    async fn read_header(
+        &self,
+        leaf_offset: u64,
+        key: &[u8],
+        buf: &mut [u8],
+    ) -> Result<Option<RecordHeader>> {
+        let buf_size = self.leaf_node_buf_size(leaf_offset);
+        let read_buf_size = self.file.read_at(&mut buf[..buf_size], leaf_offset).await?;
+        if read_buf_size != buf_size {
+            Err(anyhow!("Can't read entire leaf node"))
+        } else {
+            self.read_header_buf(&buf[..buf_size], key, self.header.record_header_size)
+                .map(|opt| opt.map(|(header, _)| header))
+        }
+    }
+
+    fn leaf_node_buf_size(&self, leaf_offset: u64) -> usize {
+        // if we read last leaf, it may be shorter
+        ((self.file_size() - leaf_offset) as usize).min(BLOCK_SIZE)
+    }
+
+    fn read_header_buf(
+        &self,
+        headers: &[u8],
+        key: &[u8],
+        record_header_size: usize,
+    ) -> Result<Option<(RecordHeader, usize)>> {
+        let mut l = 0i32;
+        let mut r: i32 = (headers.len() / record_header_size) as i32 - 1;
+        while l <= r {
+            let m = (l + r) / 2;
+            let m_off = record_header_size * m as usize;
+            let record_end = m_off + record_header_size;
+            let record_header: RecordHeader = deserialize(&headers[m_off..record_end])?;
+            match key.cmp(record_header.key()) {
```

Comment:
> Do we still need to worry about this moment if we work in-memory?

From performance point of view, we don't. But the ordering issue is the global one, because if we change the order on disk (due to different key bytes interpretation) we will need to change it here too

Reply:
Current idea is to fix ordering by adding `PearlKey` with proper array comparison: https://github.com/qoollo/pearl/issues/123

Reply:
Seems like this task is better to do in separate PR, because this one is already too big.. Resolving this conversation